File: tree/0111_minimum_depth_of_binary_tree/0111_minimum_depth_of_binary_tree.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def minDepth(self,root):
        if root is None:
            return 0 
        
        if root.left and root.right:
            return min(self.minDepth(root.left), self.minDepth(root.right)) + 1 
        else:
            return max(self.minDepth(root.left), self.minDepth(root.right)) + 1 



root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)
root.right.right= TreeNode(4)

print(Solution().minDepth(root))

File: tree/0111_minimum_depth_of_binary_tree/0111_minimum_depth_of_binary_tree.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}

	if root.Left == nil {
		return minDepth(root.Right) + 1
	}
	if root.Right == nil {
		return minDepth(root.Left) + 1
	}

	return min(minDepth(root.Left), minDepth(root.Right)) + 1
}

func min(a, b int) int {
	if a < b {
		return a
	} else {
		return b
	}
}

func main() {
	root := &TreeNode{1, nil, nil}
	root.Left = &TreeNode{2, nil, nil}
	root.Right = &TreeNode{3, nil, nil}
	root.Right.Right = &TreeNode{4, nil, nil}
	res := minDepth(root)
	fmt.Println(res)
}
